src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.CurlResponseMessage.cs

```diff
@@ -0,0 +1,500 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using size_t = System.UInt64; // TODO: IntPtr
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+        private sealed class CurlResponseMessage : HttpResponseMessage
+        {
+            internal readonly EasyRequest _easy;
+            private readonly CurlResponseStream _responseStream;
+
+            internal CurlResponseMessage(EasyRequest easy)
+            {
+                Debug.Assert(easy != null, "Expected non-null EasyRequest");
+                _easy = easy;
+
+                _responseStream = new CurlResponseStream(this);
+                RequestMessage = easy.RequestMessage;
+                Content = new StreamContent(_responseStream);
+            }
+
+            internal CurlResponseStream ResponseStream
+            {
+                get { return _responseStream; }
+            }
+        }
+
+        /// <summary>
+        /// Provides a response stream that allows libcurl to transfer data asynchronously to a reader.
+        /// When writing data to the response stream, either all or none of the data will be transferred,
+        /// and if none, libcurl will pause the connection until a reader is waiting to consume the data.
+        /// Readers consume the data via ReadAsync, which registers a read state with the stream, to be
+        /// filled in by a pending write.  Read is just a thin wrapper around ReadAsync, since the underlying
+        /// mechanism must be asynchronous to prevent blocking libcurl's processing.
+        /// </summary>
+        private sealed class CurlResponseStream : Stream
+        {
+            /// <summary>A cached task storing the Int32 value 0.</summary>
+            private static readonly Task<int> s_zeroTask = Task.FromResult(0);
+
+            /// <summary>
+            /// A sentinel object used in the <see cref="_completed"/> field to indicate that
+            /// the stream completed successfully.
+            /// </summary>
+            private static readonly Exception s_completionSentinel = new Exception("s_completionSentinel");
+
+            /// <summary>A object used to synchronize all access to state on this response stream.</summary>
+            private readonly object _lockObject = new object();
+
+            /// <summary>The response message that owns this response stream.</summary>
+            private readonly CurlResponseMessage _parentResponse;
```

Comment:
More of a doubt. We are only using the EasyRequest inside this. Fine with this for future extensibility